File: packages/format/validate.mjs
#!/usr/bin/env node
// Validate session/v0 documents against the schema plus the invariants JSON
// Schema can't express (unique span ids, acyclic parent chains). The single
// source of truth for the format's rules — the server and the CLI both call
// validateRun(). With no arguments (run directly) it checks the bundled
// examples, so "the examples validate" stays executable.
// usage: node validate.mjs [session.json ...]
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import Ajv2020Module from "ajv/dist/2020.js";
import addFormatsModule from "ajv-formats";
// Static JSON import (not a disk read) so the schema inlines when the CLI is
// bundled for distribution — a bundled/binary build has no schema file to read.
import schema from "./session.schema.json" with { type: "json" };

const Ajv2020 = Ajv2020Module.default ?? Ajv2020Module;
const addFormats = addFormatsModule.default ?? addFormatsModule;

const ajv = new Ajv2020({ allErrors: true, allowUnionTypes: true });
addFormats(ajv);
const validateSchema = ajv.compile(schema);

export function invariantErrors(run) {
  const byId = new Map();
  for (const s of run.spans) {
    if (byId.has(s.id)) return [`duplicate span id "${s.id}"`];
    byId.set(s.id, s);
  }
  const acyclic = new Set();
  for (const s of run.spans) {
    const seen = new Set();
    let cur = s;
    while (cur && !acyclic.has(cur.id)) {
      if (seen.has(cur.id)) return [`parent_id cycle involving span "${cur.id}"`];
      seen.add(cur.id);
      cur = cur.parent_id ? byId.get(cur.parent_id) : undefined;
    }
    for (const id of seen) acyclic.add(id);
  }
  return [];
}

/** (data) => string[] of errors; empty means valid. Schema + invariants. */
export function validateRun(data) {
  if (!validateSchema(data)) {
    return (validateSchema.errors ?? [])
      .slice(0, 20)
      .map((e) => `${e.instancePath || "/"} ${e.message}`);
  }
  return invariantErrors(data);
}

// Basename, not import.meta.url equality: when the CLI is bundled into one
// file every module shares import.meta.url, so URL-equality would fire this
// block on `slink <cmd>`. The entry basename is validate.mjs only when this
// file is run directly.
const isMain = path.basename(process.argv[1] ?? "") === "validate.mjs";

if (isMain) {
  const here = path.dirname(fileURLToPath(import.meta.url));
  const files = process.argv.slice(2).length
    ? process.argv.slice(2)
    : ["chat.json", "agent-eval.json"].map((f) => path.join(here, "examples", f));

  let failed = false;
  for (const file of files) {
    const data = JSON.parse(await readFile(file, "utf8"));
    const errors = validateRun(data);
    if (errors.length) {
      failed = true;
      console.error(`✗ ${path.relative(process.cwd(), file)}`);
      for (const e of errors) console.error(`    ${e}`);
    } else {
      console.log(`✓ ${path.relative(process.cwd(), file)}`);
    }
  }
  process.exit(failed ? 1 : 0);
}

File: go/internal/format/validate.go
// Package format validates session/v0 documents: the JSON Schema (embedded
// from the drift-checked copy of packages/format/session.schema.json) plus
// the invariants the schema can't express — the same rules validate.mjs
// enforces for the server and JS CLI.
package format

import (
	"bytes"
	_ "embed"
	"fmt"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

//go:embed session.schema.json
var schemaJSON []byte

var compiled = mustCompile()

func mustCompile() *jsonschema.Schema {
	doc, err := jsonschema.UnmarshalJSON(bytes.NewReader(schemaJSON))
	if err != nil {
		panic(err)
	}
	c := jsonschema.NewCompiler()
	if err := c.AddResource("session.schema.json", doc); err != nil {
		panic(err)
	}
	s, err := c.Compile("session.schema.json")
	if err != nil {
		panic(err)
	}
	return s
}

// ValidateRun mirrors validateRun: schema errors first (capped), then the
// structural invariants (unique span ids, acyclic parent chains).
func ValidateRun(data any) []string {
	if err := compiled.Validate(data); err != nil {
		if ve, ok := err.(*jsonschema.ValidationError); ok {
			errs := []string{}
			for _, cause := range flatten(ve) {
				errs = append(errs, cause)
				if len(errs) >= 20 {
					break
				}
			}
			return errs
		}
		return []string{err.Error()}
	}
	return invariantErrors(data)
}

func flatten(ve *jsonschema.ValidationError) []string {
	if len(ve.Causes) == 0 {
		loc := ve.InstanceLocation
		path := "/"
		if len(loc) > 0 {
			path = ""
			for _, p := range loc {
				path += "/" + p
			}
		}
		return []string{fmt.Sprintf("%s %s", path, ve.ErrorKind.LocalizedString(nil))}
	}
	out := []string{}
	for _, c := range ve.Causes {
		out = append(out, flatten(c)...)
	}
	return out
}

func invariantErrors(data any) []string {
	run, _ := data.(map[string]any)
	spans, _ := run["spans"].([]any)
	byID := map[string]map[string]any{}
	for _, sv := range spans {
		s, _ := sv.(map[string]any)
		id, _ := s["id"].(string)
		if _, dup := byID[id]; dup {
			return []string{fmt.Sprintf("duplicate span id %q", id)}
		}
		byID[id] = s
	}
	acyclic := map[string]bool{}
	for _, sv := range spans {
		s, _ := sv.(map[string]any)
		seen := map[string]bool{}
		cur := s
		for cur != nil {
			id, _ := cur["id"].(string)
			if acyclic[id] {
				break
			}
			if seen[id] {
				return []string{fmt.Sprintf("parent_id cycle involving span %q", id)}
			}
			seen[id] = true
			parent, _ := cur["parent_id"].(string)
			if parent == "" {
				break
			}
			cur = byID[parent]
		}
		for id := range seen {
			acyclic[id] = true
		}
	}
	return []string{}
}
